File: java/src/main/java/cash/weby/sdk/WebycashSDK.java
package cash.weby.sdk;

import com.sun.jna.*;
import com.sun.jna.ptr.PointerByReference;

public class WebycashSDK {

    public static class WebycashException extends RuntimeException {
        public final int code;
        public WebycashException(int code, String message) { super(message); this.code = code; }
    }

    private interface Lib extends Library {
        Lib INSTANCE = Native.load("webycash_sdk", Lib.class);
        int weby_wallet_open(String path, PointerByReference out);
        int weby_wallet_open_with_seed(String path, Pointer seed, long seedLen, PointerByReference out);
        void weby_wallet_free(Pointer wallet);
        int weby_wallet_balance(Pointer wallet, PointerByReference out);
        int weby_wallet_insert(Pointer wallet, String webcash);
        int weby_wallet_pay(Pointer wallet, String amount, String memo, PointerByReference out);
        int weby_wallet_check(Pointer wallet);
        int weby_wallet_merge(Pointer wallet, int maxOutputs, PointerByReference out);
        int weby_wallet_recover(Pointer wallet, String masterSecretHex, int gapLimit, PointerByReference out);
        int weby_wallet_stats(Pointer wallet, PointerByReference out);
        int weby_wallet_export_snapshot(Pointer wallet, PointerByReference out);
        int weby_wallet_encrypt_seed(Pointer wallet, String password);
        String weby_version();
        String weby_last_error_message();
        int weby_amount_parse(String amountStr, long[] out);
        int weby_amount_format(long wats, PointerByReference out);
        void weby_free_string(Pointer ptr);
    }

    private static void check(int rc) {
        if (rc != 0) {
            String msg = Lib.INSTANCE.weby_last_error_message();
            throw new WebycashException(rc, msg != null ? msg : "Error " + rc);
        }
    }

    private static String takeString(PointerByReference ref) {
        Pointer p = ref.getValue();
        if (p == null) return "";
        String s = p.getString(0);
        Lib.INSTANCE.weby_free_string(p);
        return s;
    }

    public static String version() { return Lib.INSTANCE.weby_version(); }

    public static long amountParse(String s) {
        long[] out = new long[1];
        check(Lib.INSTANCE.weby_amount_parse(s, out));
        return out[0];
    }

    public static String amountFormat(long wats) {
        PointerByReference out = new PointerByReference();
        check(Lib.INSTANCE.weby_amount_format(wats, out));
        return takeString(out);
    }

    public static class Wallet implements AutoCloseable {
        private Pointer ptr;

        public Wallet(String path) {
            PointerByReference out = new PointerByReference();
            WebycashSDK.check(Lib.INSTANCE.weby_wallet_open(path, out));
            this.ptr = out.getValue();
        }

        public Wallet(String path, byte[] seed) {
            if (seed.length != 32) throw new IllegalArgumentException("Seed must be 32 bytes");
            Memory mem = new Memory(32);
            mem.write(0, seed, 0, 32);
            PointerByReference out = new PointerByReference();
            WebycashSDK.check(Lib.INSTANCE.weby_wallet_open_with_seed(path, mem, 32, out));
            this.ptr = out.getValue();
        }

        @Override public void close() { if (ptr != null) { Lib.INSTANCE.weby_wallet_free(ptr); ptr = null; } }
        public String balance() { PointerByReference o = new PointerByReference(); WebycashSDK.check(Lib.INSTANCE.weby_wallet_balance(ptr, o)); return takeString(o); }
        public void insert(String webcash) { WebycashSDK.check(Lib.INSTANCE.weby_wallet_insert(ptr, webcash)); }
        public String pay(String amount, String memo) { PointerByReference o = new PointerByReference(); WebycashSDK.check(Lib.INSTANCE.weby_wallet_pay(ptr, amount, memo, o)); return takeString(o); }
        public void check() { WebycashSDK.check(Lib.INSTANCE.weby_wallet_check(ptr)); }
        public String merge(int maxOutputs) { PointerByReference o = new PointerByReference(); WebycashSDK.check(Lib.INSTANCE.weby_wallet_merge(ptr, maxOutputs, o)); return takeString(o); }
        public String recover(String masterSecretHex, int gapLimit) { PointerByReference o = new PointerByReference(); WebycashSDK.check(Lib.INSTANCE.weby_wallet_recover(ptr, masterSecretHex, gapLimit, o)); return takeString(o); }
        public String stats() { PointerByReference o = new PointerByReference(); WebycashSDK.check(Lib.INSTANCE.weby_wallet_stats(ptr, o)); return takeString(o); }
        public String exportSnapshot() { PointerByReference o = new PointerByReference(); WebycashSDK.check(Lib.INSTANCE.weby_wallet_export_snapshot(ptr, o)); return takeString(o); }
        public void encryptSeed(String password) { WebycashSDK.check(Lib.INSTANCE.weby_wallet_encrypt_seed(ptr, password)); }
    }
}

File: kotlin/src/main/kotlin/cash/weby/sdk/WebycashSDK.kt
package cash.weby.sdk

import com.sun.jna.*
import com.sun.jna.ptr.PointerByReference
import java.io.Closeable

private interface WebycashLib : Library {
    fun weby_wallet_open(path: String, out: PointerByReference): Int
    fun weby_wallet_open_with_seed(path: String, seed: Pointer, seedLen: Long, out: PointerByReference): Int
    fun weby_wallet_free(wallet: Pointer)
    fun weby_wallet_balance(wallet: Pointer, out: PointerByReference): Int
    fun weby_wallet_insert(wallet: Pointer, webcash: String): Int
    fun weby_wallet_pay(wallet: Pointer, amount: String, memo: String?, out: PointerByReference): Int
    fun weby_wallet_check(wallet: Pointer): Int
    fun weby_wallet_merge(wallet: Pointer, maxOutputs: Int, out: PointerByReference): Int
    fun weby_wallet_recover(wallet: Pointer, masterSecretHex: String, gapLimit: Int, out: PointerByReference): Int
    fun weby_wallet_stats(wallet: Pointer, out: PointerByReference): Int
    fun weby_wallet_export_snapshot(wallet: Pointer, out: PointerByReference): Int
    fun weby_wallet_encrypt_seed(wallet: Pointer, password: String): Int
    fun weby_version(): String
    fun weby_last_error_message(): String?
    fun weby_amount_parse(amountStr: String, out: LongArray): Int
    fun weby_amount_format(wats: Long, out: PointerByReference): Int
    fun weby_free_string(ptr: Pointer)

    companion object {
        val INSTANCE: WebycashLib = Native.load("webycash_sdk", WebycashLib::class.java)
    }
}

class WebycashException(val code: Int, message: String) : RuntimeException(message)

private fun check(rc: Int) {
    if (rc != 0) throw WebycashException(rc, WebycashLib.INSTANCE.weby_last_error_message() ?: "Error $rc")
}

private fun takeString(ref: PointerByReference): String {
    val p = ref.value ?: return ""
    val s = p.getString(0)
    WebycashLib.INSTANCE.weby_free_string(p)
    return s
}

fun version(): String = WebycashLib.INSTANCE.weby_version()

fun amountParse(s: String): Long {
    val out = LongArray(1)
    check(WebycashLib.INSTANCE.weby_amount_parse(s, out))
    return out[0]
}

fun amountFormat(wats: Long): String {
    val out = PointerByReference()
    check(WebycashLib.INSTANCE.weby_amount_format(wats, out))
    return takeString(out)
}

class Wallet private constructor(private var ptr: Pointer?) : Closeable {

    companion object {
        fun open(path: String): Wallet {
            val out = PointerByReference()
            check(WebycashLib.INSTANCE.weby_wallet_open(path, out))
            return Wallet(out.value)
        }

        fun openWithSeed(path: String, seed: ByteArray): Wallet {
            require(seed.size == 32) { "Seed must be 32 bytes" }
            val mem = Memory(32).also { it.write(0, seed, 0, 32) }
            val out = PointerByReference()
            check(WebycashLib.INSTANCE.weby_wallet_open_with_seed(path, mem, 32, out))
            return Wallet(out.value)
        }
    }

    override fun close() { ptr?.let { WebycashLib.INSTANCE.weby_wallet_free(it); ptr = null } }

    fun balance(): String { val o = PointerByReference(); check(WebycashLib.INSTANCE.weby_wallet_balance(ptr!!, o)); return takeString(o) }
    fun insert(webcash: String) { check(WebycashLib.INSTANCE.weby_wallet_insert(ptr!!, webcash)) }
    fun pay(amount: String, memo: String = ""): String { val o = PointerByReference(); check(WebycashLib.INSTANCE.weby_wallet_pay(ptr!!, amount, memo, o)); return takeString(o) }
    fun check() { cash.weby.sdk.check(WebycashLib.INSTANCE.weby_wallet_check(ptr!!)) }
    fun merge(maxOutputs: Int = 20): String { val o = PointerByReference(); cash.weby.sdk.check(WebycashLib.INSTANCE.weby_wallet_merge(ptr!!, maxOutputs, o)); return takeString(o) }
    fun recover(masterSecretHex: String, gapLimit: Int = 20): String { val o = PointerByReference(); cash.weby.sdk.check(WebycashLib.INSTANCE.weby_wallet_recover(ptr!!, masterSecretHex, gapLimit, o)); return takeString(o) }
    fun stats(): String { val o = PointerByReference(); cash.weby.sdk.check(WebycashLib.INSTANCE.weby_wallet_stats(ptr!!, o)); return takeString(o) }
    fun exportSnapshot(): String { val o = PointerByReference(); cash.weby.sdk.check(WebycashLib.INSTANCE.weby_wallet_export_snapshot(ptr!!, o)); return takeString(o) }
    fun encryptSeed(password: String) { cash.weby.sdk.check(WebycashLib.INSTANCE.weby_wallet_encrypt_seed(ptr!!, password)) }
}
